### PR TITLE
Add htslib-backed VCF/BCF reader

### PR DESCRIPTION
**Pull request by riasc**
### Added
- `vcf_reader` (`include/genogrove/io/vcf_reader.hpp`, `src/io/vcf_reader.cpp`): an htslib-backed `file_reader<vcf_entry>` mirroring `bam_reader`. Parses sites (CHROM, POS, ID, REF/ALT, QUAL, FILTER, typed INFO) plus FORMAT and per-sample genotype columns. Coordinates are 0-based half-open (`start = POS - 1`, `end = start + len(REF)`) to match the other readers; convert via `gdt::interval(start, end - 1)`.
- `vcf_entry` with `sample_genotype` (decoded GT alleles, phasing, `gt_string()`, `is_hom_ref()`) and convenience predicates (`passed_filter()`, `is_snp()`, `is_indel()`). Options: `vcf_reader_options{parse_info, parse_samples, skip_filtered}` with `defaults()` / `sites_only()` factories.
- 24-case GoogleTest suite (`tests/io/vcffile-test.cpp`) plus fixtures (`test.vcf`, `test_rich.vcf`, `test.vcf.gz`) covering typed INFO/FORMAT (int/float/string/flag, multi-valued), multi-allelic records, genotype decoding, missing QUAL/ID/GT, options, bgzip, and the standard reader edge cases (file-not-found, empty file, line counter, has_next, iterator, move ctor/assignment).

## QC
- [x] I, as a human being, have checked each line of code in this pull request
- [x] `genogrove_io_tests` builds and all `VcfReaderTest.*` cases pass
- [x] VCF POS is decoded 0-based (POS 100 -> `start == 99`); REF `AT` deletion yields `end == start + 2`
- [x] Multi-allelic `A>G,T` yields `alt.size() == 2`; missing `QUAL` (`.`) sets `qual_missing == true`
- [x] Phased `1|1` and missing `./.` genotypes decode via `gt_string()`; `skip_filtered` drops non-PASS records
- [x] bgzipped `.vcf.gz` is detected as VCF/GZIP and reads identically to the plain file

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a VCF/BCF reader that auto-detects input type, reads variant records with 0-based coordinates, REF/ALT, QUAL, FILTER, typed INFO, and per-sample FORMAT/genotypes
  * Genotype utilities and predicates (phasing, homozygous-reference, GT string) plus options for sites-only parsing and skipping filtered records

* **Tests**
  * Added comprehensive tests and VCF fixtures covering headers, multi-allelic records, INFO/FORMAT parsing, iteration, edge cases, compression, and move semantics

* **Documentation**
  * Updated changelog describing the new reader and options

### CODE REVIEW

File: src/io/vcf_reader.cpp [hunk 1]
```diff
@@ -0,0 +1,443 @@
...
+
+        // REF / ALT. d.allele[0] is REF; [1..n_allele) are ALT alleles.
+        entry.ref.clear();
+        entry.alt.clear();
+        if (rec->n_allele > 0) {
+            entry.ref = rec->d.allele[0];
+            entry.alt.reserve(rec->n_allele - 1);
+            for (int i = 1; i < rec->n_allele; ++i) {
+                entry.alt.emplace_back(rec->d.allele[i]);
+            }
```
[the remark this answers is not in the document]
Comment (riasc) [reply src/io/vcf_reader.cpp:246]: **Verdict: Address (partially).** Verified against htslib 1.23. The `"."` half of this is already correct — htslib collapses a monomorphic `ALT=.` to `n_allele==1` (REF only), so `entry.alt` is already empty (confirmed: a record with `ALT=.` parses to `alt.size()==0`). The `<*>` half is valid: symbolic alleles (`<*>`, `<NON_REF>`) are stored as real alleles, and `is_indel()` then misclassifies them because it compares string lengths (`"<*>".size() != ref.size()` → reports an indel). The fix isn't to drop `<*>` (it's meaningful gVCF data) but to (1) correct the doc comment's false "empty for `<*>`" promise and (2) make `is_snp()`/`is_indel()` ignore symbolic alleles (those starting with `<` or equal to `*`). See the summary comment for the consolidated finding.

File: src/io/vcf_reader.cpp [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+                    if (n > 0) {
+                        entry.info[tag] = std::vector<int32_t>(int_buf, int_buf + n);
+                    }
+                    break;
+                }
+
+                case BCF_HT_REAL: {
+                    int n = bcf_get_info_float(header, rec, tag, &float_buf, &n_float_buf);
+                    if (n > 0) {
+                        entry.info[tag] = std::vector<float>(float_buf, float_buf + n);
```
[the remark this answers is not in the document]
Comment (riasc) [reply src/io/vcf_reader.cpp:305]: **Verdict: Address.** Confirmed real bug. A haploid sample on a `Number=R` FORMAT field at a biallelic site parses to `AD=[7, -2147483647]` — `bcf_int32_vector_end` (and `*_missing`) sentinels leak into the published vector because `parse_samples_fields` copies the flat padded buffer verbatim with a fixed `per = n / n_samples` stride. The existing `test_rich.vcf` masked this since every value was present. Same exposure on variable-length numeric INFO. Fix: trim trailing `bcf_int32_vector_end`/`bcf_float_vector_end` per sample and decide on a representation for interior `*_missing` elements. See summary comment.

File: tests/io/vcffile-test.cpp [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+    fs::path test_data_dir;
+    fs::path vcf_path;
+    fs::path rich_path;
+    fs::path gz_path;
+
+    void SetUp() override {
+        test_data_dir = fs::current_path() / "io" / "data";
+        vcf_path = test_data_dir / "test.vcf";
+        rich_path = test_data_dir / "test_rich.vcf";
+        gz_path = test_data_dir / "test.vcf.gz";
```
[the remark this answers is not in the document]
Comment (riasc) [reply tests/io/vcffile-test.cpp:37]: **Verdict: Skip.** This matches the established repo convention — every io test resolves fixtures via `fs::current_path() / "io" / "data"` (see `bamfile-test.cpp`, `bedfile-test.cpp`, `gfffile-test.cpp`). CTest sets the working directory so this resolves correctly in CI and locally. Diverging in only the new file would be inconsistent; if we want source-relative fixture roots that should be a separate change across the whole io suite.

File: tests/io/vcffile-test.cpp [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+    gio::vcf_entry entry;
+    EXPECT_FALSE(reader.read_next(entry));
+    EXPECT_FALSE(reader.has_next());
+
+    std::vector<gio::vcf_entry> entries(reader.begin(), reader.end());
+    EXPECT_TRUE(entries.empty());
+    EXPECT_TRUE(reader.get_error_message().empty());
+
+    fs::remove(empty_vcf);
+}
```
[the remark this answers is not in the document]
Comment (riasc) [reply tests/io/vcffile-test.cpp:392]: **Verdict: Skip.** This mirrors the existing `BamReaderTest.EmptySamFileHeaderOnly`, which writes `test_data_dir / "test_empty_records.sam"` the same way and removes it at the end. GoogleTest runs cases sequentially within a binary, so there's no intra-binary race, and the suites are separate executables. Keeping the pattern consistent with the existing reader tests is preferable to a one-off divergence here.

**Comment by riasc** [comment]
## Code Review

Reviewed the diff for correctness and consistency with the existing io readers. Two real correctness issues, both confirmed empirically against htslib 1.23 with throwaway fixtures.

### Bug — htslib padding sentinels leak into numeric FORMAT/INFO values
`parse_samples_fields` slices the flat `bcf_get_format_*` buffer with a fixed `per = n / n_samples` stride and copies it verbatim. For ragged numeric fields (e.g. `Number=R`/`G`/`.` with samples of differing ploidy), htslib right-pads each sample to the max width with `bcf_int32_vector_end` (`-2147483647`) and uses `bcf_int32_missing`/`bcf_float_missing` for absent values. A haploid sample on a biallelic `AD` field parses to `[7, -2147483647]` — the sentinel is published as a real depth. `test_rich.vcf` didn't catch it because every value was present.
**Fix:** strip trailing `*_vector_end` per sample before storing, and pick a representation for interior `*_missing` elements. Add a mixed-ploidy / partially-missing fixture to lock it in. (Same root cause CodeRabbit flagged inline.)

### Bug — symbolic alleles misclassified, and a false doc promise
`vcf_entry::alt` is documented as "empty for monomorphic / `<*>`". The monomorphic `ALT=.` case is already fine (htslib collapses it to `n_allele==1`, so `alt` is empty). But symbolic alleles (`<*>`, `<NON_REF>`) are stored as real alleles, and `is_indel()` then reports them as indels because it length-compares `"<*>"` against REF. 
**Fix:** correct the doc comment (don't promise emptiness for `<*>`; it's kept verbatim as it's meaningful gVCF data) and have `is_snp()`/`is_indel()` skip symbolic alleles (leading `<` or `*`).

### Suggestion (optional)
- `read_next` treats every `bcf_read` return of `0` as success without checking `record->errcode`. htslib returns `0` with `errcode` set on non-critical issues (e.g. an undefined header tag). The BAM reader is similarly lenient, so this is consistent — noting only in case stricter validation is wanted later.

Test paths via `fs::current_path()` and the inline temp-file pattern both match the existing io reader tests, so I'd keep them as-is (see inline replies).

The rest — Rule of Five / move semantics, htslib handle and `bcf_get_format_string` buffer cleanup, the 0-based coordinate convention, error-message contract, and the 24-case suite structure — all look correct and consistent with `bam_reader`.

**Comment by riasc** [comment]
## Codex Review

_Independent automated review by OpenAI Codex._

Skipped this run — Codex usage limit reached (resets Jul 2). Re-run `/review-pr` after that to include it.

File: src/io/vcf_reader.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (riasc) [reply src/io/vcf_reader.cpp:305]: Addressed in 222027e: per-sample numeric FORMAT vectors now stop at the first `bcf_int32_vector_end` / `bcf_float_is_vector_end` sentinel, so ragged fields like `AD` no longer leak padding. Added `test_edge.vcf` + `RaggedFormatTrimsVectorEndPadding` (a haploid sample on a biallelic `AD` now yields `[7]`, not `[7, -2147483647]`).

File: src/io/vcf_reader.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (riasc) [reply src/io/vcf_reader.cpp:246]: Addressed in 222027e: corrected the `vcf_entry::alt` doc comment, and `is_snp()` / `is_indel()` now skip symbolic alleles via `is_symbolic_allele()` (leading `<`, `*`, `.`, or breakend `[`/`]`). `<NON_REF>` is kept verbatim as real gVCF data but no longer misclassified as an indel. Added `SymbolicAlleleNotClassifiedAsIndel` and `MonomorphicAltYieldsEmptyAlt` tests.
